lib/src/barback/dartdevc/scratch_space.dart

```diff
@@ -10,32 +10,69 @@ import 'package:path/path.dart' as p;
 
 import '../../io.dart';
 
+typedef Future<Asset> AssetReader(AssetId id);
+
 /// An on-disk temporary environment for running executables that don't have
 /// a standard Dart library API.
 class ScratchSpace {
   final Directory tempDir;
   final Directory packagesDir;
+  final AssetReader getAsset;
+
+  // Assets which have a file created but it is still being written to.
+  final _pendingWrites = <AssetId, Future>{};
 
-  ScratchSpace._(Directory tempDir)
+  ScratchSpace._(Directory tempDir, this.getAsset)
       : packagesDir = new Directory(p.join(tempDir.path, 'packages')),
         this.tempDir = tempDir;
 
-  /// Creates a new [ScratchSpace] containing [assetIds].
+  factory ScratchSpace(Future<Asset> getAsset(AssetId id)) {
+    var tempDir = new Directory(createSystemTempDir());
+    return new ScratchSpace._(tempDir, getAsset);
+  }
+
+  /// Copies [assetIds] to [tempDir] if they don't exist.
   ///
   /// Any [Asset] that is under a `lib` dir will be output under a `packages`
   /// directory corresponding to its package, and any other assets are output
   /// directly under the temp dir using their unmodified path.
-  static Future<ScratchSpace> create(
-      Iterable<AssetId> assetIds, Stream<List<int>> readAsset(AssetId)) async {
-    var tempDir = new Directory(createSystemTempDir());
+  Future ensureAssets(Iterable<AssetId> assetIds) async {
     var futures = <Future>[];
     for (var id in assetIds) {
-      var filePath = p.join(tempDir.path, _relativePathFor(id));
-      ensureDir(p.dirname(filePath));
-      futures.add(createFileFromStream(readAsset(id), filePath));
+      var file = fileFor(id);
+      if (file.existsSync()) {
+        var pending = _pendingWrites[id];
+        if (pending != null) futures.add(pending);
+      } else {
+        file.createSync(recursive: true);
+        var done = () async {
+          var asset = await getAsset(id);
+          await createFileFromStream(asset.read(), file.path);
+          _pendingWrites.remove(id);
+        }();
+        _pendingWrites[id] = done;
+        futures.add(done);
+      }
+    }
+    return Future.wait(futures);
+  }
+
+  /// Deletes all files for [package] from the temp dir (synchronously).
+  ///
+  /// This always deletes the [package] dir under [packagesDir].
+  ///
+  /// If [isRootPackage] then this also deletes all top level entities under
+  /// [tempDir] other than the [packagesDir].
+  void deletePackageFiles(String package, bool isRootPackage) {
```

Comment:
It's probably worth making `isRootPackage` named.

https://www.dartlang.org/guides/language/effective-dart/design#avoid-positional-boolean-parameters